flink-core/src/main/java/org/apache/flink/core/fs/FileSystem.java

```diff
@@ -440,8 +467,21 @@ public static FileSystem getUnguardedFileSystem(final URI fsUri) throws IOExcept
 				try (TemporaryClassLoaderContext ignored = TemporaryClassLoaderContext.of(classLoader)) {
 					fs = factory.create(uri);
 				}
-			}
-			else {
+			} else if (!ALLOWED_FALLBACK_FILESYSTEMS.contains(uri.getScheme()) &&
+					DIRECTLY_SUPPORTED_FILESYSTEM.containsKey(uri.getScheme())) {
+				final Collection<String> plugins = DIRECTLY_SUPPORTED_FILESYSTEM.get(uri.getScheme());
+				throw new UnsupportedFileSystemSchemeException(String.format(
+					"Could not find a file system implementation for scheme '%s'. The scheme is " +
+						"directly supported by Flink through the following plugin%s: %s. Please ensure that each " +
+						"plugin resides within its own subfolder within the plugins directory. See https://ci.apache" +
+						".org/projects/flink/flink-docs-stable/ops/plugins.html for more information. If you want to " +
+						"use a Hadoop file system for that scheme, please add the scheme to the configuration fs" +
+						".allowed-fallback-filesystems.",
+						uri.getScheme(),
+						plugins.size() == 1 ? "" : "s",
+						String.join(", ", plugins)
+					));
```

Comment:
I'm not sure if I like the `DIRECTLY_SUPPORTED_FILESYSTEM` list, since if it drifts out of sync, it can start decreasing the usability.

Maybe as a way to mitigate this issue, we should add a reference to https://ci.apache.org/projects/flink/flink-docs-stable/ops/filesystems/ for further/more precise  informations?

Reply:
Added reference.